www/widgets/rt.js

```diff
@@ -164,55 +166,90 @@ class ssl {
+        return wrap_class(class {
+            constructor() {
+                this.elt = rt_constants.ROOT_CANVAS;
+                this.bindings = {}
+                this.event_names_to_key = {}
+                rt_constants.WINDOW = this;
+            }
+
+            bind(key, callback) {
+                let event_name, fn;
+
+                if (['<Up>', '<Down>', '<Left>', '<Right>'].indexOf(key) !== -1) {
+                    event_name = "keydown";
+                    fn = function(e) {
+                        if (e.key == 'Arrow' + key.substr(1, key.length-2)) {
+                            e.preventDefault();
+                            callback({}).catch(on_error);
+                        }
                     };
-                });
-            } else if (key.length == 1) {
-                window.addEventListener("keydown", function(e) {
-                    if (e.key == key) {
-                        e.preventDefault();
-                        fn({ char: e.key }).catch(on_error);
-                    }
-                });
-            } else {
-                console.error("Trying to bind unsupported event", key);
+                } else if (key == '<Return>') {
+                    event_name = "keydown";
+                    fn = function(e) {
+                        if (e.key == 'Enter') {
+                            e.preventDefault();
+                            callback({}).catch(on_error);
+                        }
+                    };
+                } else if (['<Button-1>', '<Button-2>', '<Button-3>'].indexOf(key) !== -1) {
+                    event_name = "mousedown";
+                    fn = function(e) {
+                        if (e.button == key.substr(8, 1) - 1) {
+                            callback({ x: e.offsetX, y: e.offsetY }).catch(on_error);
+                        }
+                    };
+                } else if (key == '<Configure>') {
+                    event_name = "resize";
+                    fn = function(e) {
+                        callback({}).catch(on_error);
+                    };
+                } else if (key == '<Key>') {
+                    event_name = "keydown";
+                    fn = function(e) {
+                        if (e.key.length == 1) {
+                            e.preventDefault();
+                            callback({ char: e.key }).catch(on_error);
+                        };
+                    };
+                } else if (key.length == 1) {
+                    event_name = "keydown";
+                    fn = function(e) {
+                        if (e.key == key) {
+                            e.preventDefault();
+                            callback({ char: e.key }).catch(on_error);
+                        }
+                    };
+                } else {
+                    console.error("Trying to bind unsupported event", key);
+                }
+                if (key in this.bindings) {
```

Comment:
Here is where I avoid re-binding. It changes the bound function and then returns.